src/buffer_tree.cpp

```diff
@@ -339,45 +343,50 @@ flush_ret_t inline BufferTree::flush_control_block(BufferControlBlock *bcb) {
 	}
 }
 
-// load data from buffer memory location so long as the key matches
-// what we expect
-data_ret_t BufferTree::get_data(work_t task) {
-	data_ret_t data;
-	Node key = task.first;
-	data.first = key;
-	File_Pointer off = 0;
-	BufferControlBlock *bcb = buffers[task.second];
+// ask the buffer tree for data
+// this function may sleep until data is available
+bool BufferTree::get_data(data_ret_t &data) {
+	File_Pointer idx = 0;
 
-	// printf("getting data from positon %u and for key %u\n", task.second, key);
+	// make a request to the circular buffer for data
+	std::pair<int, queue_elm> queue_data;
+	bool got_data = cq->peek(queue_data);
 
-	char *serial_data = (char *) malloc(bcb->size());
-	int len = pread(backing_store, serial_data, bcb->size(), bcb->offset());
-	// printf("read %lu bytes\n", len);
-	if (len == -1) {
-		printf("ERROR get_data failed to read from buffer %i, %s\n", bcb->get_id(), strerror(errno));
-		return data;
-	}
+	if (!got_data)
+		return false; // we got no data so return not valid
 
-	while(off < (uint64_t)len) {
-		update_t upd = deserialize_update(serial_data + off);
-		// printf("got update: %u %u %i\n", upd.first.first, upd.first.second, upd.second);
+	int i         = queue_data.first;
+	queue_elm elm = queue_data.second;
+	char *serial_data = elm.data;
+	uint32_t len      = elm.size;
+
+	if (len == 0)
+		return false; // we got no data so return not valid
+
+	data.second.clear(); // remove any old data from the vector
+	uint32_t vec_len  = len / serial_update_size;
+	data.second.reserve(vec_len); // reserve space for our updates
+
+	// assume the first key is correct so extract it
+	Node key = load_key(serial_data);
+	data.first = key;
+
+	while(idx < (uint64_t) len) {
+		update_t upd = deserialize_update(serial_data + idx);
+		// printf("got update: %lu %lu\n", upd.first, upd.second);
 		if (upd.first == 0 && upd.second == 0) {
-			break; // got a null entry so clear that out
+			break; // got a null entry so done
 		}
 
 		if (upd.first == key) {
```

Comment:
is there a possibility this won't be `key`?

Reply:
I'm not sure exactly what you mean. There is a possibility that the key will be 0 but we don't consider self loops. So a null entry is 0 and 0.

Reply:
is there a case (if it's not a null entry) that `upd.first != key`?

Reply:
Oh gotcha. The rest of the code should prevent that from happening. I have these checks of upd.first just to make sure and to detect the null entry.

Reply:
right but if upd.first isn't key, we just lose the update then no?

Reply:
Correct, in my head it was better to miss these updates than let them go to the wrong node. I thought the code complexity of placing it where it belonged wasn't worth it considering it really shouldn't happen.

Reply:
but missing updates is unacceptable, since it might give us the wrong answer to a graph query

Reply:
Good point. This really shouldn't happen though so I don't think worth the hassle of writing the code to handle it. However, I think it's a bad idea to just quietly drop it as I'm currently doing. Instead I'll have the system throw an error.